File: packages/schemas/src/jober_schemas/recovery.py
from typing import Any
from uuid import UUID

from jober_schemas.common import SchemaBase


class SelfAssessmentRead(SchemaBase):
    attempt_index: int
    strategy_name: str
    failure_class: str
    tried: str
    happened: str
    next_change: str


class FailureReportRead(SchemaBase):
    job_target_id: UUID
    company: str
    role: str
    apply_url: str | None
    failed_step: str
    failure_class: str
    inferred_reason: str
    recommended_manual_action: str
    safe_to_retry: bool
    attempt_count: int
    screenshot_object_key: str | None = None
    trace_object_key: str | None = None
    dom_snapshot_object_key: str | None = None
    attempted_actions: list[str] = []
    self_assessments: list[SelfAssessmentRead] = []


class FailureAnalyticsBucket(SchemaBase):
    platform: str
    failure_class: str
    count: int
    circuit_tripped: bool = False


class FailureAnalyticsRead(SchemaBase):
    buckets: list[FailureAnalyticsBucket]
    alerts: list[dict[str, Any]] = []

File: apps/web/src/lib/api/recovery.ts
import { ApiError, apiFetch } from "@/lib/api/client";

export interface FailureReport {
  job_target_id: string;
  company: string;
  role: string;
  apply_url?: string | null;
  failed_step: string;
  failure_class: string;
  inferred_reason: string;
  recommended_manual_action: string;
  safe_to_retry: boolean;
  attempt_count: number;
  screenshot_object_key?: string | null;
  self_assessments?: Array<{
    attempt_index: number;
    strategy_name: string;
    tried: string;
    happened: string;
    next_change: string;
  }>;
}

export interface FailureAnalytics {
  buckets: Array<{
    platform: string;
    failure_class: string;
    count: number;
    circuit_tripped: boolean;
  }>;
  alerts: Array<Record<string, unknown>>;
}

export async function fetchFailureReportForJob(
  jobTargetId: string,
): Promise<FailureReport | null> {
  try {
    return await apiFetch<FailureReport>(`/api/job-targets/${jobTargetId}/failure-report`);
  } catch (err: unknown) {
    if (err instanceof ApiError && err.status === 404) {
      return null;
    }
    throw err;
  }
}

export async function fetchFailureAnalytics(): Promise<FailureAnalytics> {
  return apiFetch<FailureAnalytics>("/api/recovery/failure-analytics");
}
